src/Bicep.Core/Analyzers/Linter/Rules/InterpolateNotConcatRule.cs

```diff
@@ -37,22 +38,32 @@ private class Visitor : SyntaxVisitor
             public List<IBicepAnalyzerDiagnostic> diagnostics = new List<IBicepAnalyzerDiagnostic>();
 
             private InterpolateNotConcatRule parent;
+            private SemanticModel model;
 
-            public Visitor(InterpolateNotConcatRule parent)
+            public Visitor(InterpolateNotConcatRule parent, SemanticModel model)
             {
                 this.parent = parent;
+                this.model = model;
             }
 
             public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
             {
-                if (syntax.NameEquals("concat"))
+                if (syntax.NameEquals("concat") && !syntax.GetParseDiagnostics().Any())
                 {
-                    if (CreateFix(syntax) is CodeFix fix)
+                    // We should only suggest rewriting concat() calls that result in a string (concat can also operate on and
+                    // return arrays)
+                    var resultType = this.model.GetTypeInfo(syntax);
+                    if (!(resultType is AnyType) && TypeValidator.AreTypesAssignable(resultType, LanguageConstants.String))
```

Comment:
[nit] latest version of C# allows negative pattern matching with `resultType is not AnyType`

Reply:
Much better, thanks!